### PR TITLE
Adds support for compact De Bruijn graph extraction

### PR DESCRIPTION
**Pull request by ctb**
Adds hashtable/graph functions "find_high_degree_nodes" and "traverse_linear_path", along with script "sandbox/extract-compact-dbg.py".

Extracted from MinHash branch (#1351).

Needed before review:

- [x] some tests for the extract-compact-dbg.py script.

----

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [x] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the `ChangeLog`?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)
- [x] Is the Copyright year up to date?

### CODE REVIEW

**Comment by ctb** [comment]
Note to @luizirber - I copied the sandbox/extract-compact-dbg.py script over from walk-dbg.py in https://github.com/spacegraphcats/spacegraphcats/, and removed the MinHash dependencies. It should be a good place to experiment with compact DBG.  I also refactored some of the hashtable.cc high_degree_node stuff so that it no longer uses an internal C++ SeenSet.

File: khmer/_khmer.cc [hunk 1]
```diff
@@ -1614,8 +1681,8 @@ hashtable_find_all_tags_list(khmer_KHashtable_Object * me, PyObject * args)
 
     Py_END_ALLOW_THREADS
 
-        PyObject * x = (PyObject *) create_HashSet_Object(tags,
-                                                          hashtable->ksize());
+    PyObject * x = (PyObject *) create_HashSet_Object(tags,
```
Comment (ctb) [inline khmer/_khmer.cc]: (this is a minor formatting fix that we should leave in)

File: lib/kmer_hash.cc [hunk 2]
```diff
@@ -210,7 +211,7 @@ KmerIterator::KmerIterator(const char * seq,
     _nbits_sub_1 = (_ksize*2 - 2);
 
     index = _ksize - 1;
-    length = strlen(seq);
+    length = strlen(_seq);
```
Comment (ctb) [inline lib/kmer_hash.cc]: use the internal _seq rather than the passed-in one. probably no big deal.

**Comment by ctb** [comment]
Ready for review, @standage!

File: lib/hashtable.cc [hunk 3]
```diff
@@ -997,4 +997,79 @@ void Hashtable::get_kmer_counts(const std::string &s,
...
+                                             SeenSet &adjacencies,
+                                             SeenSet &visited, Hashtable &bf,
+                                             SeenSet &high_degree_nodes)
+    const
+{
+    unsigned int size = 0;
+
+    auto filter = [&] (Kmer& n) -> bool {
+        return true;
+    };
```
Comment (standage) [inline lib/hashtable.cc]: With my primarily ANSI C background I'm not sure how to parse this. Pointers? (no pun intended)

Comment (ctb) [inline lib/hashtable.cc]: This is a lambda function that (in this case) always returns True.

However, I removed the need for it in the traverse(...) function so it
shouldn't be here anyway :).  (See traversal.cc, traverse_right(...).)

File: sandbox/extract-compact-dbg.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+from __future__ import print_function
+import khmer
+import screed
+import argparse
+from collections import OrderedDict
+import sys
+
+# graph settings
+DEFAULT_KSIZE=31
+NODEGRAPH_SIZE=8e8 # small, big is 2e8
```
Comment (standage) [inline sandbox/extract-compact-dbg.py]: I'm not sure how 8e8 corresponds to "small" while 2e8 corresponds to "big".

File: sandbox/extract-compact-dbg.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('seqfiles', nargs='+')
+    parser.add_argument('-o', '--output', default=None)
+    parser.add_argument('-k', '--ksize', default=DEFAULT_KSIZE, type=int)
+    parser.add_argument('-x', '--tablesize', default=NODEGRAPH_SIZE,
+                            type=float)
+    parser.add_argument('--force', action='store_true')
+    #parser.add_argument('--gml', action='store_true')
```
Comment (standage) [inline sandbox/extract-compact-dbg.py]: GML is default, no? Is this option necessary?

**Comment by standage** [comment]
LGTM, aside from the posted comments.

File: sandbox/extract-compact-dbg.py (hunk 5 above)
Comment (ctb) [inline sandbox/extract-compact-dbg.py]: Not any more, no. And it was commented out but should still be removed ;)

**Comment by ctb** [comment]
addressed!

**Comment by ctb** [comment]
w00t thanks!

**Comment by standage** [comment]
👍

**Pull request merged**